meilisearch-lib/src/index_resolver/mod.rs

```diff
@@ -81,18 +71,53 @@ impl std::ops::Deref for IndexUid {
 }
 
 impl TryInto<IndexUid> for String {
-    type Error = IndexResolverError;
+    type Error = IndexUidFormatError;
+
+    fn try_into(self) -> std::result::Result<IndexUid, IndexUidFormatError> {
+        IndexUid::from_str(&self)
+    }
+}
+
+#[derive(Debug)]
+pub struct IndexUidFormatError {
+    invalid_uid: String,
+}
 
-    fn try_into(self) -> Result<IndexUid> {
-        IndexUid::new(self)
+impl fmt::Display for IndexUidFormatError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "invalid index uid `{}`, the uid must be an integer \
+            or a string containing only alphanumeric characters \
+            a-z A-Z 0-9, hyphens - and underscores _.",
+            self.invalid_uid,
+        )
+    }
+}
+
+impl Error for IndexUidFormatError {}
+
+impl From<IndexUidFormatError> for IndexResolverError {
+    fn from(error: IndexUidFormatError) -> Self {
+        Self::BadlyFormatted(error.invalid_uid)
     }
 }
 
 impl FromStr for IndexUid {
-    type Err = IndexResolverError;
+    type Err = IndexUidFormatError;
 
-    fn from_str(s: &str) -> Result<IndexUid> {
-        IndexUid::new(s.to_string())
+    fn from_str(uid: &str) -> std::result::Result<IndexUid, IndexUidFormatError> {
+        if !uid
+            .chars()
+            .all(|x| x.is_ascii_alphanumeric() || x == '-' || x == '_')
+            || !(1..=400).contains(&uid.len())
```

Comment:
I understand that this was not the scope of this PR, but I find that this line is not easy to understand, could you change it to something like:

```suggestion
            || uid.len() < 1 || uid.len() > 400
```

Reply:
done! Clippy suggested `uid.is_empty()` instead of `uid.len() < 1` so I used that but let me know if you want me to change it